### PR TITLE
ext_proc: revise service api for attributes

### PR DESCRIPTION
**Pull request by jbohanon**
Commit Message: Revise service API to reflect changes discussed in https://github.com/envoyproxy/envoy/issues/32125
Additional Description:
Risk Level: Low - feature merged very recently and has not been released
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

**Comment by jbohanon** [comment]
cc @htuch @rshriram

**Comment by wbpcode** [comment]
https://github.com/envoyproxy/envoy/pull/30747

**Review by wbpcode — COMMENTED** [review]
Thanks for you contribution. I taken a look to the dicussion and you previous PR. (But it's a little complex so I am not sure I get it correctly, feel free to point it out if I am wrong.)

I think if you want to revert (or partly revert)previous implementation, you should add the `[#not-implemented-hide:]` back to the `request_attributes` and `response_attributes` because you may want to change the behavior in the implementation.

Actually, if we haven't figured out a final solution or API, we should just revert previous two implementation commits directly. It would be more clean. cc @htuch

**Comment by rshriram** [comment]
@wbpcode I dont think we need to revert any changes from previous PRs. What is being done here is to add the attributes to the top of processing request, commit the implementation and _then_ deprecate the request/response level ones. The implementation should require only minor changes (populate ProcessingRequest.attributes instead of ProcessingRequest.request|response_headers.attributes) on top of the code in #31090.

File: api/envoy/service/ext_proc/v3/external_processor.proto [hunk 1]
```diff
@@ -110,8 +110,19 @@ message ProcessingRequest {
     HttpTrailers response_trailers = 7;
   }
 
+  // [#not-implemented-hide:]
+  // TODO(jbohanon) reserve field as part of rework detailed in https://github.com/envoyproxy/envoy/issues/32125
```
Comment (rshriram) [inline api/envoy/service/ext_proc/v3/external_processor.proto:114]: @htuch , per discussion in the issue thread, shall we just keep this field as it is, maintaining a clear separation between attributes derived from the connection/request/resp vs those pushed through config or internal state?

File: api/envoy/service/ext_proc/v3/external_processor.proto [hunk 2]
```diff
@@ -110,8 +110,19 @@ message ProcessingRequest {
...
   // Dynamic metadata associated with the request.
   config.core.v3.Metadata metadata_context = 8;
+
+  // [#not-implemented-hide:]
+  // The values of properties selected by the ``request_attributes``
+  // or ``response_attributes`` list in the configuration. Each entry
+  // in the list is populated from the standard
+  // :ref:`attributes <arch_overview_attributes>` supported across Envoy.
+  // This field also contains dynamic metadata from forwarding namespaces
```
Comment (rshriram) [inline api/envoy/service/ext_proc/v3/external_processor.proto:123]: Suggest removing this line following comment above

**Comment by wbpcode** [comment]
> @wbpcode I dont think we need to revert any changes from previous PRs. If we agree on just moving the attributes outside the HttpHeaders message, then this API change becomes much simpler and the implementation should require only minor changes (populate ProcessingRequest.attributes instead of ProcessingRequest.request|response_headers.attributes) on top of the code in #31090. 

Yeah. But the premise is we can agree on with all the API changes (attributes and metadata context) and complete the code and test updates in the windows. (Today may the last day?)

And from my perspective, this PR seems more like a partly reverting rather than refactoring, for now.

Considering the time windows, I still suggest you revert the previous commits and we can merge the reverting directly. Then you can create a new PR for these features immediately.

Or we must act very quickly.

**Comment by wbpcode** [comment]
Another solution is defer the `metadata_contexr` related changes to another PR, because it was merged at one week ago and still has one week time to update it.

Then this PR could only update the attributes implementation (like the @rshriram 's suggestion) and API which would LGTM to me.

**Comment by jbohanon** [comment]
I have updated this to be more tightly-scoped to the attributes fields. I agree that reverting the committed code is not necessary. The discussion around the field `metadata_context` is still ongoing in the issue thread, but moving `attributes` up to be a first-class field in the `ProcessingRequest` message is not controversial.

**Comment by wbpcode** [comment]
@jbohanon LGTM to the API, and please update the implementation.

Here is the timeline:

2024/01/24: merged the feature of the `request_attributes` and `response_attributes`. This attributes will be add to the `attributes` in the headers field.

Now, if you want to move the `attributes` to the `ProcessingRequest`, you should complete all refactoring (include API and implementation) before 2024/02/07.

File: api/envoy/service/ext_proc/v3/external_processor.proto [hunk 3]
```diff
@@ -204,6 +211,7 @@ message HttpHeaders {
   config.core.v3.HeaderMap headers = 1;
 
   // [#not-implemented-hide:]
+  // TODO(jbohanon) reserve field as part of rework detailed in https://github.com/envoyproxy/envoy/issues/32125
```
Comment (wbpcode) [inline api/envoy/service/ext_proc/v3/external_processor.proto:214]: Is this TODO necessary? If the `attributes` in the `ProcessingRequet` should be used, then please mark this field as deprecated directly.

Comment (jbohanon) [reply api/envoy/service/ext_proc/v3/external_processor.proto:214]: TODO will be removed at time of implementation

**Comment by jbohanon** [comment]
> Now, if you want to move the attributes to the ProcessingRequest, you should complete all refactoring (include API and implementation) before 2024/02/07.

Sounds good. For my education, where is the two-week contract specified?

**Comment by jbohanon** [comment]
Possible implementation put up here: https://github.com/jbohanon/envoy/pull/1

I can pull those changes into this PR if desired, but I still need to work on the tests

**Comment by wbpcode** [comment]
> > Now, if you want to move the attributes to the ProcessingRequest, you should complete all refactoring (include API and implementation) before 2024/02/07.
> 
> Sounds good. For my education, where is the two-week contract specified?

https://github.com/envoyproxy/envoy/blob/main/api%2FAPI_VERSIONING.md

From this API guiding, we can only do a break change to exist API in two weeks after it is merged or it's marked as WIP.

Now you want to abort the `attributes` in the header message directly, IMO, it is like a break change. 🤔

**Comment by wbpcode** [comment]
I also noticed that the `attributes` is existing for a long time and still be marked as `not-implement-hide`.
So, I am not sure if we can treat this as special case. 🤔

**Comment by rshriram** [comment]
@wbpcode that is precisely the reason we are doing this change now and trying to remove the HttpHeaders.attributes because that field was never implemented till now. So we have the ability to make this breaking change.

**Review by rshriram — APPROVED** [review]
LGTM
Thank you so much for acting quickly.

**Review by htuch — APPROVED** [review]
LGTM. To be clear, the headers `attributes` field was never marked not `[#not-implemented-hide:]` (it's unchanged on line 206 above). So, from a stable API perspective it was never a stable field :)

**Comment by jbohanon** [comment]
Thanks @htuch. I am working through the integration test updates now and will ping again once everything is green on CI. 

There was an outstanding question on https://github.com/jbohanon/envoy/pull/1 re: simply reserving this field that was never marked as stable/implemented. What do you think of that?

**Review by wbpcode — COMMENTED** [review]
LGTM to the code. Could you check the integration tests? Seems that the error may is related to this PR.

**Comment by jbohanon** [comment]
@htuch @wbpcode Thanks for the reviews. I have fixed up the integration test and am ready for another :eyes:

**Review by wbpcode — APPROVED** [review]
LGTM. Thanks.

**Comment by rshriram** [comment]
@jbohanon one question. Our API says request_attributes and response_attributes. What about the connection.* attributes documented here (https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/advanced/attributes) ? Does the code reject those?

**Comment by rshriram** [comment]
Put another way, should we just have a generic attributes repeated field in ext_proc filter instead of request and response attributes that does not seem to add any real value, as the wire proto has all attributes in one bag.

**Comment by jbohanon** [comment]
> @jbohanon one question. Our API says request_attributes and response_attributes. What about the connection.* attributes documented here (https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/advanced/attributes) ? Does the code reject those?

@rshriram the qualifier in front of `_attributes` specifies when that list is sent, not what attribute prefixes are accepted. It is the case that `connection.*` attributes are available on the request path, but it is not the case that `response.*` attributes are available on the request path. Keeping them separated allows users to list sensible attributes, and just those which they need, for the different request and response processing requests.

**Comment by rshriram** [comment]
so if I set BOTH request_attributes and response attributes in ext_proc filter to the following, 
request.host, source.address, response.code

what happens during request headers/body events and response headers/body events?

Can I assume that there will be exactly ONe copy of request.host, source.address in ProcessingRequest.attributes for request headers/body calls and exactly ONE copy of request.host, source.address, response.code in ProcessingRequest.attributes for response headers/body calls?

**Comment by jbohanon** [comment]
> so if I set BOTH request_attributes and response attributes in ext_proc filter to the following, request.host, source.address, response.code
> 
> what happens during request headers/body events and response headers/body events?
> 
> Can I assume that there will be exactly ONe copy of request.host, source.address in ProcessingRequest.attributes for request headers/body calls and exactly ONE copy of request.host, source.address, response.code in ProcessingRequest.attributes for response headers/body calls?

Correct, given:
```proto
ExternalProcessor:
  request_attributes:
    - request.host
    - source.address
    - response.code
  response_attributes:
    - request.host
    - source.address
    - response.code
```

There will be exactly ONE copy of `request.host` and `source.address`, and exactly ZERO copies of `response.code` in the `ProcessingRequest.attributes` for request headers/body/trailers processing calls. There will be exactly ONE copy of `request.host` and `source.address`, and `response.code` in the `ProcessingRequest.attributes` for response headers/body/trailers processing calls.

**Review by rshriram — APPROVED** [review]

**Review by yanavlasov — APPROVED** [review]

File: source/extensions/filters/http/ext_proc/ext_proc.cc [hunk 4]
```diff
@@ -771,6 +757,21 @@ void Filter::addDynamicMetadata(const ProcessorState& state, ProcessingRequest&
...
 }
 
+void Filter::addAttributes(ProcessorState& state, ProcessingRequest& req) {
+  if (!state.sendAttributes(config_->expressionManager())) {
+    return;
+  }
+
+  auto activation_ptr = Filters::Common::Expr::createActivation(
+      &config_->expressionManager().localInfo(), state.callbacks()->streamInfo(),
+      state.requestHeaders(), dynamic_cast<const Http::ResponseHeaderMap*>(state.responseHeaders()),
```
Comment (tyxia) [inline source/extensions/filters/http/ext_proc/ext_proc.cc:767]: @jbohanon 
when you pass the response header to createActivation, is there guarantee that it won't be mutated? 
const RequestOrResponseHeaderMap* ptr can still be pointed to a different location?

**Pull request merged**